src/ralph/lib/custom_fields/fields.py

```diff
@@ -0,0 +1,352 @@
+import operator
+from collections import defaultdict
+from functools import reduce
+
+from django.contrib.contenttypes.fields import (
+    create_generic_related_manager,
+    GenericRelation,
+    ReverseGenericRelatedObjectsDescriptor
+)
+from django.contrib.contenttypes.models import ContentType
+from django.db import connection, models
+
+from ralph.admin.helpers import get_field_by_relation_path, getattr_dunder
+
+
+class CustomFieldsWithInheritanceRelation(GenericRelation):
+    """
+    Relation for custom fields which handle inheritance of custom fields values
+    from dependent models (fields) specified in custom_fields_inheritance.
+
+    To use inheritance of custom fields, define `custom_fields_inheritance`
+    in your model, specifying paths of fields from which `CustomFieldValue`s
+    should be inherited. `custom_fields_inheritance` supports dunder convention
+    for nested fields.
+
+    Example:
+    class Restaurant(models.Model):
+        city = models.ForeignKey(City)
+
+        custom_fields = CustomFieldsWithInheritanceRelation(CustomFieldValue)
+        custom_fields_inheritance = ['city', 'city__country']
+    """
+    def contribute_to_class(self, cls, name, **kwargs):
+        super().contribute_to_class(cls, name, **kwargs)
+        # use ReverseGenericRelatedObjectsWithInheritanceDescriptor
+        # instead of ReverseGenericRelatedObjectsDescriptor
+        setattr(
+            cls,
+            self.name,
+            ReverseGenericRelatedObjectsWithInheritanceDescriptor(
+                self,
+                self.for_concrete_model
+            )
+        )
+
+
+def _prioritize_custom_field_values(objects, model, content_type):
+    """
+    Sort custom field values by priorities and leave the ones with
+    biggest priority for each custom field type.
+
+    Priority is defined as follows:
+    * the biggest priority has custom field defined directly for
+      instance
+    * then next priority has CFV from first field on
+      `custom_fields_inheritance` list, then second from this list and
+      so on
+    """
+    ct_priority = [content_type.id]
+    for field_path in model.custom_fields_inheritance:
+        field = get_field_by_relation_path(model, field_path)
+        content_type = ContentType.objects.get_for_model(field.rel.to)
+        ct_priority.append(content_type.id)
+    ct_priority = {
```

Comment:
it will NOT work - notice that in index is enumerate is value of dict item, not key